File: process.c
//
//  process.c
//  elenbot
//
//  Created by Jascha Burmeister on 11.02.19.
//  Copyright © 2019 Jascha Burmeister. All rights reserved.
//

#include "process.h"
#include <mach/mach.h>
#import <sys/proc_info.h>
#import <libproc.h>

int get_process_ids(pid_t* pids, int* size) {
    int numberOfProcesses = proc_listpids(PROC_ALL_PIDS, 0, NULL, 0);
    pid_t _pids[numberOfProcesses];
    proc_listpids(PROC_ALL_PIDS, 0, pids, (int) sizeof(_pids));
    *size = numberOfProcesses;
    return 0;
}

int get_process_name(pid_t pid, char* buffer) {
    int numberOfProcesses = proc_listpids(PROC_ALL_PIDS, 0, NULL, 0);
    pid_t pids[numberOfProcesses];
    bzero(pids, sizeof(pids));
    proc_listpids(PROC_ALL_PIDS, 0, pids, (int) sizeof(pids));
    for (int i = 0; i < numberOfProcesses; ++i) {
        
        if(pids[i] == pid) {
            proc_name(pids[i], buffer, 1024);
            return 0;
        }
        
    }
    
    return 0;
}

int get_remote_task(int pid, mach_port_t* task) {
    kern_return_t kern_return = task_for_pid(mach_task_self(), pid, task);
    if (kern_return != KERN_SUCCESS)
    {
        printf("task_for_pid() failed, error %d - %s\n", kern_return, mach_error_string(kern_return));
        return kern_return;
    }
    return 0;
}

int get_base_address(int pid, unsigned long long* base_address) {
    
    kern_return_t kern_return;
    mach_port_t task;
    
    kern_return = task_for_pid(mach_task_self(), pid, &task);
    if (kern_return != KERN_SUCCESS)
    {
        printf("task_for_pid() failed, error %d - %s\n", kern_return, mach_error_string(kern_return));
        return kern_return;
    }
    
    kern_return_t kret;
    vm_region_basic_info_data_t info;
    vm_size_t size;
    mach_port_t object_name;
    mach_msg_type_number_t count;
    vm_address_t firstRegionBegin = 0;
    vm_address_t lastRegionEnd = 0;
    vm_size_t fullSize = 0;
    count = VM_REGION_BASIC_INFO_COUNT_64;
    mach_vm_address_t address = 1;
    int regionCount = 0;
    int flag = 0;
    while (flag == 0)
    {
        //Attempts to get the region info for given task
        kret = mach_vm_region(task, &address, (mach_vm_size_t *) &size, VM_REGION_BASIC_INFO, (vm_region_info_t) &info, &count, &object_name);
        if (kret == KERN_SUCCESS)
        {
            if (regionCount == 0)
            {
                firstRegionBegin = address;
                regionCount += 1;
            }
            fullSize += size;
            address += size;
        }
        else
            flag = 1;
    }
    lastRegionEnd = address;
    //    printf("Base Address: %lu\n",firstRegionBegin);
    //    printf("lastRegionEnd: %lu\n",lastRegionEnd);
    //    printf("fullSize: %lu\n",fullSize);
    
    
    *base_address = firstRegionBegin;
    
    return 0;

}

int get_image_symbol_address(mach_port_t task, const char *image_name, const char *symbol_name, unsigned long long* address) {
    *address = lorgnette_lookup_image(task, symbol_name, image_name);

    if(*address == 0) {
        return 1;
    } else {
        return 0;
    }

}

File: Process.swift
//
//  Process.swift
//  elenbot
//
//  Created by Jascha Burmeister on 19.02.19.
//  Copyright © 2019 Jascha Burmeister. All rights reserved.
//

import Foundation

enum ProcessError: Error {
    case getRemoteTaskFailed
    case getBaseAddressFailed
    case getProcessNameByPidFailed
    case getProcessIdsFailed
    case getImageSymbolFailed
}

class Process {
    
    private let client: Client
    public private(set) var remoteTask : mach_port_t
    public private(set) var baseAddress : CUnsignedLongLong
    
    init(client: Client) throws {
        self.client = client
        self.remoteTask = try Process.getRemoteTask(client.pid)
        self.baseAddress = try Process.getBaseAddress(client.pid)
    }
    
    static func getProcessIds() throws -> Array<pid_t> {
        var buffer = [pid_t](repeating: 0, count: 5000)
        var size : Int32  = 0
        if(get_process_ids(&buffer, &size) != 0) {
            throw ProcessError.getProcessIdsFailed
        }
        let arr  = convert(length: Int(size), data: &buffer, pid_t.self)
        return arr
    }
    
    static func getProcessNameByPid(_ pid : pid_t) throws -> String {
        var buffer = [CChar](repeating: 0, count: 1024)
        if(get_process_name(pid, &buffer) != 0) {
            throw ProcessError.getProcessNameByPidFailed
        }

        return String(cString: buffer)
    }
    
    static func getRemoteTask(_ pid : pid_t) throws -> mach_port_t {
        var task: mach_port_t = 0
        if(get_remote_task(pid, &task) != 0) {
            throw ProcessError.getRemoteTaskFailed
        }
        return task
    }
    
    static func getBaseAddress(_ pid : pid_t) throws -> UInt64  {
        var address: UInt64 = 0x0
        if(get_base_address(pid, &address) != 0) {
            throw ProcessError.getBaseAddressFailed
        }
        return address
    }
    
    func getImageSymbolAddress(imageName: String, symbolName: String) throws -> UInt64 {
        var address: UInt64 = 0x0
        if(get_image_symbol_address(remoteTask, imageName, symbolName, &address) != 0) {
            throw ProcessError.getImageSymbolFailed
        }
        return address
    }
    
    private static func convert<T>(length: Int, data: UnsafeMutablePointer<pid_t>, _: T.Type) -> [T] {
        let numItems = length/MemoryLayout<T>.stride
        let buffer = data.withMemoryRebound(to: T.self, capacity: numItems) {
            UnsafeBufferPointer(start: $0, count: numItems)
        }
        return Array(buffer)
    }
    
    
}
